hpy/devel/src/runtime/ctx_type.c

```diff
@@ -931,17 +1023,60 @@ ctx_Type_GenericNew(HPyContext *ctx, HPy h_type, HPy *args, HPy_ssize_t nargs, H
 }
 
 _HPy_HIDDEN void*
-ctx_AsStruct(HPyContext *ctx, HPy h)
+ctx_AsStruct_Object(HPyContext *ctx, HPy h)
 {
-    return _HPy_PyObject_Payload(_h2py(h));
+    return _HPy_Payload(_h2py(h), HPyType_BuiltinShape_Object);
 }
 
 _HPy_HIDDEN void*
-ctx_AsStructLegacy(HPyContext *ctx, HPy h)
+ctx_AsStruct_Legacy(HPyContext *ctx, HPy h)
 {
     return _h2py(h);
 }
 
+_HPy_HIDDEN void*
+ctx_AsStruct_Type(HPyContext *ctx, HPy h)
+{
+    return _HPy_Payload(_h2py(h), HPyType_BuiltinShape_Type);
```

Comment:
btw. I verified (by disassembling in the debugger) that those functions really just do the offset computation if compiled with `-O3`:
```
(lldb) di -f
mytest.cpython-39-darwin.so`ctx_AsStruct_Object:
    0x1073497a0 <+0>: pushq  %rbp
    0x1073497a1 <+1>: movq   %rsp, %rbp
->  0x1073497a4 <+4>: leaq   0x10(%rsi), %rax
    0x1073497a8 <+8>: popq   %rbp
    0x1073497a9 <+9>: retq
    ```